### PR TITLE
Change how fail_hard transactions are handled.

### PR DESCRIPTION
**Pull request by undertome**
FIXES: #2847

* Transactions that are submitted with the fail_hard flag
  and that result in any TER code besides tesSUCCESS shall
  be neither queued nor held.

### CODE REVIEW

File: src/test/jtx/Env_test.cpp [hunk 1]
```diff
@@ -348,6 +349,79 @@ class Env_test : public beast::unit_test::suite
...
+
+        BEAST_EXPECT(jr[jss::result][jss::engine_result] == "temBAD_FEE");
+        BEAST_EXPECT(env.app().getTxQ().getMetrics(*env.current()).txCount == queueTxCount);
+        BEAST_EXPECT(env.app().getOPs().getLocalTxCount () == localTxCnt);
+
+        jr = applyTxn(offer(alice, XRP(1000), USD(1000)));
+
+        BEAST_EXPECT(jr[jss::result][jss::engine_result] == "tecUNFUNDED_OFFER");
+        BEAST_EXPECT(env.app().getTxQ().getMetrics(*env.current()).txCount == queueTxCount);
+        BEAST_EXPECT(env.app().getOPs().getLocalTxCount () == localTxCnt);
```
Comment (ximinez) [inline src/test/jtx/Env_test.cpp]: This test case is a duplicate.

File: src/ripple/app/misc/NetworkOPs.cpp [hunk 2]
```diff
@@ -1126,8 +1129,9 @@ void NetworkOPsImp::apply (std::unique_lock<std::mutex>& batchLock)
             }
             else if (e.result == terQUEUED)
             {
-                JLOG(m_journal.debug()) << "Transaction is likely to claim a " <<
-                    "fee, but is queued until fee drops";
+                JLOG(m_journal.debug()) << "Transaction is likely to claim a "
+                                        << "fee, but is queued until fee drops";
```
[the remark this answers is not in the document]
Comment (undertome) [reply src/ripple/app/misc/NetworkOPs.cpp]: Done

File: src/ripple/app/misc/NetworkOPs.cpp [hunk 3]
```diff
@@ -1156,16 +1156,18 @@ void NetworkOPsImp::apply (std::unique_lock<std::mutex>& batchLock)
                 e.transaction->setStatus (INVALID);
             }
 
-            if (addLocal)
+            auto const enforceFailHard = e.failType == FailHard::yes && e.result != tesSUCCESS;
```
[the remark this answers is not in the document]
Comment (undertome) [reply src/ripple/app/misc/NetworkOPs.cpp]: Done

File: src/test/jtx/Env_test.cpp (hunk 1 above)
Comment (undertome) [reply src/test/jtx/Env_test.cpp]: Done

File: src/ripple/app/misc/NetworkOPs.cpp [hunk 4]
```diff
@@ -1055,7 +1055,10 @@ void NetworkOPsImp::apply (std::unique_lock<std::mutex>& batchLock)
                     // we check before adding to the batch
                     ApplyFlags flags = tapNONE;
                     if (e.admin)
-                        flags = flags | tapUNLIMITED;
+                        flags |= tapUNLIMITED;
+
+                    if (e.local && e.failType == FailHard::yes)
+                        flags |= tapFAIL_HARD;
```
[the remark this answers is not in the document]
Comment (undertome) [reply src/ripple/app/misc/NetworkOPs.cpp]: Done

File: src/test/jtx/Env_test.cpp [hunk 5]
```diff
@@ -348,6 +349,79 @@ class Env_test : public beast::unit_test::suite
...
+        using namespace jtx;
+        Env env(*this);
+        auto const gw = Account("gateway");
+        auto const USD = gw["USD"];
+
+        auto const alice = Account {"alice"};
+        env.fund (XRP(10000), alice);
+
+        auto const localTxCnt = env.app().getOPs().getLocalTxCount();
+        auto const queueTxCount = env.app().getTxQ().getMetrics(*env.current()).txCount;
```
[the remark this answers is not in the document]
Comment (undertome) [reply src/test/jtx/Env_test.cpp]: Thanks for the sample commit - done

File: src/test/jtx/Env_test.cpp [hunk 6]
```diff
@@ -348,6 +349,79 @@ class Env_test : public beast::unit_test::suite
...
+        jr = applyTxn(offer(alice, XRP(1000), USD(1000)));
+
+        BEAST_EXPECT(jr[jss::result][jss::engine_result] == "tecUNFUNDED_OFFER");
+        BEAST_EXPECT(env.app().getTxQ().getMetrics(*env.current()).txCount == queueTxCount);
+        BEAST_EXPECT(env.app().getOPs().getLocalTxCount () == localTxCnt);
+
+        jr = applyTxn(noop (alice));
+
+        BEAST_EXPECT(jr[jss::result][jss::engine_result] == "tesSUCCESS");
+        BEAST_EXPECT(env.app().getOPs().getLocalTxCount () > localTxCnt);
```
[the remark this answers is not in the document]
Comment (undertome) [reply src/test/jtx/Env_test.cpp]: Done

File: src/ripple/app/misc/NetworkOPs.cpp [hunk 7]
```diff
@@ -1136,11 +1142,7 @@ void NetworkOPsImp::apply (std::unique_lock<std::mutex>& batchLock)
             }
             else if (isTerRetry (e.result))
             {
-                if (e.failType == FailHard::yes)
-                {
-                    addLocal = false;
-                }
-                else
+                if (e.failType != FailHard::yes)
```
Comment (mellery451) [inline src/ripple/app/misc/NetworkOPs.cpp]: seems like `addLocal` is a direct alias for `e.local` now (never modified), so we can just get rid of it.

File: src/ripple/app/tx/impl/Transactor.cpp [hunk 8]
```diff
@@ -670,7 +670,15 @@ Transactor::operator()()
...
+    if (isTecClaim(result) && (view().flags() & tapFAIL_HARD))
+    {
+        // If the tapFAIL_HARD flag is set, a tec result
+        // must not do anything
+
+        ctx_.discard();
+        applied = false;
+    }
+    else if ((result == tecOVERSIZE) || (result == tecKILLED) ||
         (isTecClaimHardFail (result, view().flags())))
```
Comment (mellery451) [inline src/ripple/app/tx/impl/Transactor.cpp]: strictly from a naming perspective, now that you've added a new flag called `FAIL_HARD`, having a method with the name "HardFail" in it that doesn't actually consider that flag seems misleading - just food for thought.

**Comment by intelliot** [comment]
@mDuo13 Could we write docs for how this change will be communicated to developers? Ideally we would do this prior to merging the PR. Happy to help with this if I can.

File: src/ripple/app/misc/NetworkOPs.cpp (hunk 7 above)
Comment (undertome) [reply src/ripple/app/misc/NetworkOPs.cpp]: Sounds good, done.

File: src/ripple/app/tx/impl/Transactor.cpp (hunk 8 above)
Comment (undertome) [reply src/ripple/app/tx/impl/Transactor.cpp]: Added the check for tapFAIL_HARD.

**Pull request closed without merging**